sky/backends/backend.py

```diff
@@ -45,20 +45,45 @@ def check_resources_fit_cluster(self, handle: _ResourceHandleType,
     @timeline.event
     @usage_lib.messages.usage.update_runtime('provision')
     def provision(
-            self,
-            task: 'task_lib.Task',
-            to_provision: Optional['resources.Resources'],
-            dryrun: bool,
-            stream_logs: bool,
-            cluster_name: Optional[str] = None,
-            retry_until_up: bool = False) -> Optional[_ResourceHandleType]:
+        self,
+        task: 'task_lib.Task',
+        to_provision: Optional['resources.Resources'],
+        dryrun: bool,
+        stream_logs: bool,
+        cluster_name: Optional[str] = None,
+        retry_until_up: bool = False,
+        skip_if_no_cluster_updates: bool = False,
+    ) -> Optional[_ResourceHandleType]:
+        """Provisions resources for the given task.
+
+        Args:
+            task: The task to provision resources for.
+            to_provision: Resource config to provision. Should only be None if
+                cluster_name refers to an existing cluster, whose resources will
+                be used.
```

Comment:
This does not necessarily need to be True.
```suggestion
            to_provision: Resource config to provision.
```

Reply:
Hmm... if to_provision is None and cluster_name refers to a non-existent cluster, what do we even do? We have no resource information then.

Reply:
Optimizer runs before this:
in `_execute`
https://github.com/skypilot-org/skypilot/blob/51a7e177d99fdfe73a89c04dddc385940a97a37d/sky/execution.py#L277-L279
then later
https://github.com/skypilot-org/skypilot/blob/51a7e177d99fdfe73a89c04dddc385940a97a37d/sky/execution.py#L292-L297

Inside provision, we call `_check_existing_cluster` to get the resources to provision:
https://github.com/skypilot-org/skypilot/blob/51a7e177d99fdfe73a89c04dddc385940a97a37d/sky/backends/cloud_vm_ray_backend.py#L2736-L2742
`_check_existing_cluster` assumes that if `to_provision` is None, the cluster already exists. 
https://github.com/skypilot-org/skypilot/blob/51a7e177d99fdfe73a89c04dddc385940a97a37d/sky/backends/cloud_vm_ray_backend.py#L4306-L4319
The basic logic of `_check_existing_cluster` looks like this:
- `record = backend_utils.refresh_cluster_record(cluster_name, ...)`
- if `record` is set, return provision config matching existing cluster
- if `to_provision` is None (and `record` is None, since we would have already returned), then assume that the cluster was downed and throw an exception
- otherwise return ToProvisionConfig for a new cluster based on to_provision

Reply:
`_check_existing_cluster` is one of the first calls in `_provision`. There is no optimize call between the start of `_provision` and `_check_existing_cluster`.
https://github.com/skypilot-org/skypilot/blob/51a7e177d99fdfe73a89c04dddc385940a97a37d/sky/backends/cloud_vm_ray_backend.py#L2699-L2742